src/backend/sgx/attestation.rs

```diff
@@ -0,0 +1,365 @@
+                    Some(s) => s,
+                    None => {
+                        return Err(Error::new(
+                            ErrorKind::Other,
+                            "no key size provided for setting Init Quote Ex request",
+                        ));
+                    }
+                };
+                let mut msg = Request_InitQuoteExRequest::new();
+                msg.set_timeout(TIMEOUT);
+                msg.set_b_pub_key_id(true);
+                msg.set_att_key_id(akid);
+                msg.set_buf_size(size as u64);
+                req.set_initQuoteExReq(msg);
+                Ok(req)
+            }
+            ReqType::KeySize => {
+                let akid = match akid {
+                    Some(a) => a,
+                    None => {
+                        return Err(Error::new(
+                            ErrorKind::Other,
+                            "no attestation key ID provided for setting Init Quote Ex request to get key size",
+                        ));
+                    }
+                };
+                let mut msg = Request_InitQuoteExRequest::new();
+                msg.set_timeout(TIMEOUT);
+                msg.set_b_pub_key_id(false);
+                msg.set_att_key_id(akid);
+                req.set_initQuoteExReq(msg);
+                Ok(req)
+            }
+            ReqType::Quote => {
+                let report = match report {
+                    Some(r) => r,
+                    None => {
+                        return Err(Error::new(
+                            ErrorKind::Other,
+                            "no Report provided for setting Get Quote Ex request",
+                        ));
+                    }
+                };
+                let akid = match akid {
+                    Some(a) => a,
+                    None => {
+                        return Err(Error::new(
+                            ErrorKind::Other,
+                            "no attestation key ID provided for setting Get Quote Ex request",
+                        ));
+                    }
+                };
+                let mut msg = Request_GetQuoteExRequest::new();
+                msg.set_timeout(TIMEOUT);
+                msg.set_report(report.to_vec());
+                msg.set_att_key_id(akid);
+                msg.set_buf_size(SGX_QUOTE_SIZE as u32);
+                req.set_getQuoteExReq(msg);
+                Ok(req)
+            }
+        }
+    }
+
+    fn send_request(&self, req: Request, mut stream: UnixStream) -> Result<Response, Error> {
+        // Set up writer
+        let mut buf_wrtr = vec![0u8; size_of::<u32>()];
+        match req.write_to_writer(&mut buf_wrtr) {
+            Ok(_) => {}
+            Err(e) => {
+                return Err(Error::new(
+                    ErrorKind::Other,
+                    format!("invalid protobuf Request: {:?}. Error: {:?}", req, e),
+                ));
+            }
+        }
```

Comment:
This is copied code from https://github.com/enarx/enarx/blob/7f691c91783c841c807d7fc72c8038a2210d8007/src/backend/sgx/attestation.rs
don't blame this code to @jarkkojs

Reply:
```
        // Set up writer
        req.write_to_writer(&mut buf_wrtr).ok_or_else(|| {
            Error::new(
                ErrorKind::Other,
                format!("Invalid protobuf request: {:?}", req),
            )
        })?;
```

This will give compilation error:

```
   --> src/backend/sgx/attestation.rs:111:44
    |
111 |         req.write_to_writer(&mut buf_wrtr).ok_or_else(|| {
    |                                            ^^^^^^^^^^ help: there is an associated function with a similar name: `map_or_else`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `enarx` due to previous error
```

Reply:
Now that all validation is fixed in `set_request()` can we close this? Not the prettiest code but it is still **working** code.

Reply:
Can't we do something like this?
```suggestion
       req.write_to_writer(&mut buf_wrtr).map_err(|e| Error::new(
                    ErrorKind::Other,
                    format!("invalid protobuf Request: {:?}. Error: {:?}", req, e),
                ))?;
```